pkg/controller/perconaservermongodb/smart.go

```diff
@@ -0,0 +1,180 @@
+	if cr.Spec.UpdateStrategy != api.SmartUpdateStatefulSetStrategyType {
+		return nil
+	}
+
+	if sfs.Status.UpdatedReplicas >= sfs.Status.Replicas {
+		return nil
+	}
+
+	if ok, err := cr.VersionGreaterThanOrEqual("1.4.0"); !ok || err != nil {
+		return errors.Wrap(err, "failed to compare version")
+	}
+
+	log.Info("statefullSet was changed, run smart update")
+
+	if sfs.Status.ReadyReplicas < sfs.Status.Replicas {
+		log.Info("can't start/continue 'SmartUpdate': waiting for all replicas are ready")
+		return nil
+	}
+
+	ok, err := r.isBackupRunning(cr)
+	if err != nil {
+		return fmt.Errorf("failed to check active backups: %v", err)
+	}
+	if ok {
+		log.Info("can't start 'SmartUpdate': waiting for running backups finished")
+		return nil
+	}
+
+	list := corev1.PodList{}
+	if err := r.client.List(context.TODO(),
+		&list,
+		&client.ListOptions{
+			Namespace: cr.Namespace,
+			LabelSelector: labels.SelectorFromSet(map[string]string{
+				"app.kubernetes.io/name":       "percona-server-mongodb",
+				"app.kubernetes.io/instance":   cr.Name,
+				"app.kubernetes.io/replset":    replset.Name,
+				"app.kubernetes.io/managed-by": "percona-server-mongodb-operator",
+				"app.kubernetes.io/part-of":    "percona-server-mongodb",
+			}),
+		},
+	); err != nil {
+		return fmt.Errorf("get pod list: %v", err)
+	}
+
+	primary, err := r.getPrimaryPod(cr, replset, list, secret)
+	if err != nil {
+		return fmt.Errorf("get primary pod: %v", err)
+	}
+
+	log.Info(fmt.Sprintf("primary pod is %s", primary))
+
+	waitLimit := int(replset.LivenessProbe.InitialDelaySeconds)
+
+	sort.Slice(list.Items, func(i, j int) bool {
+		return list.Items[i].Name > list.Items[j].Name
+	})
+
+	var primaryPod corev1.Pod
+	for _, pod := range list.Items {
+		pod := pod
+		if strings.HasPrefix(primary, fmt.Sprintf("%s.%s.%s", pod.Name, sfs.Name, sfs.Namespace)) {
+			primaryPod = pod
+		} else {
+			log.Info(fmt.Sprintf("apply changes to secondary pod %s", pod.Name))
+			if err := r.applyNWait(cr, sfs.Status.UpdateRevision, &pod, waitLimit); err != nil {
+				return fmt.Errorf("failed to apply changes: %v", err)
+			}
+		}
+	}
+
+	log.Info(fmt.Sprintf("apply changes to primary pod %s", primaryPod.Name))
+	if err := r.applyNWait(cr, sfs.Status.UpdateRevision, &primaryPod, waitLimit); err != nil {
+		return fmt.Errorf("failed to apply changes: %v", err)
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaServerMongoDB) isBackupRunning(cr *api.PerconaServerMongoDB) (bool, error) {
+	bcps := api.PerconaServerMongoDBBackupList{}
+	if err := r.client.List(context.TODO(), &bcps, &client.ListOptions{Namespace: cr.Namespace}); err != nil {
+		if k8sErrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, errors.Wrap(err, "get backup list")
+	}
+
+	for _, bcp := range bcps.Items {
+		if bcp.Status.State == api.BackupStateRunning ||
+			bcp.Status.State == api.BackupStateWaiting ||
+			bcp.Status.State == api.BackupStateRequested {
```

Comment:
can we check status via successful or failed state?
reason - we can add more statuses in future 
```suggestion
		if b.Status.State != api.BackupStateReady && b.Status.State != api.BackupStateError && {
```